terrat_runner/work_exec.py

```diff
@@ -139,14 +139,34 @@ def _mask_secrets(secrets, unmasked, value):
         return value
 
 
+def _extract_secrets(run_time, value):
+    if isinstance(value, str):
+        return run_time.extract_secrets(value)
+    elif isinstance(value, dict):
+        ret = []
+        for v in value.values():
+            ret.extend(_extract_secrets(run_time, v))
+        return ret
+    elif isinstance(value, list):
+        ret = []
+        for v in value:
+            ret.extend(_extract_secrets(run_time, v))
+        return ret
```

Comment:
`isinstance` accepts tuples as a second argument as well:

```suggestion
    elif isinstance(value, (list, dict)):
        values = value.values() if isinstance(value, dict) else value
        ret = []
        for v in values:
            ret.extend(_extract_secrets(run_time, v))
        return ret
```

although maybe a more maintainable way to refactor this in the meantime is with pattern matching (depends on which version of Python the action is using):

```python
def _extract_secrets(run_time, value):
    ret = []
    match value:
        case str():
            return run_time.extract_secrets(value)
        case dict():
            for v in values:
                ret.extend(_extract_secrets(run_time, v))
            return ret
        case list():
            for v in values:
                ret.extend(_extract_secrets(run_time, v))
            return ret
        case _:
            return []
```

Reply:
IMHO, combining the `isinstance` of list and dict is less readable.  We end up having to have even more ifs.  Perhaps an alternative would be to make another function that handles lists and then call that from dict.  I think this code is small enough that there isn't a huge benefit to that refactoring.  And I want to rewrite all of this in the relatively near future anyways.

This particular part is a total hack because this secrets management should be happening closer to the commands being executed rather than right before we push the results back to the server.  Which is why we end up duplicating some logic here.